pkg/sql/catalog/multiregion/region_config.go

```diff
@@ -153,28 +153,79 @@ func (r *RegionConfig) ZoneConfigExtensions() descpb.ZoneConfigExtensions {
+			numVoters = *zc.NumVoters
+		}
+		if zc.NumReplicas != nil {
+			numReplicas = *zc.NumReplicas
+		}
 		zc = extendZoneCfg(zc, *ext)
+		// TODO(janexing): to ensure that the zone config extension won't break the
+		// survival goal, here we only add restriction on the number of voters and
+		// replicas. We may want to consider adding constraint to their distribution
+		// across zones/regions as well.
+		if zc.NumVoters != nil && *zc.NumVoters < numVoters {
+			return zonepb.ZoneConfig{}, errors.Newf("zone config extension "+
+				"cannot set a voter number %v that is lower than the one required for the "+
+				"survival goal: %v with goal %v\n", *zc.NumVoters, numVoters, r.SurvivalGoal())
+		}
+		if zc.NumReplicas != nil && *zc.NumReplicas < numReplicas {
+			return zonepb.ZoneConfig{}, errors.Newf("zone config extension "+
+				"cannot set a replica number %v that is lower than the one required for the "+
+				"survival goal: %v with goal %v\n", *zc.NumReplicas, numReplicas, r.SurvivalGoal())
+		}
 	}
-	return zc
+	return zc, nil
 }
 
-// ApplyZoneConfigExtensionForRegionalIn applies the regional table zone
+// ExtendZoneConfigWithRegionalIn applies the regional table zone
 // configuration extensions for the provided region to the provided zone
 // configuration, returning the updated config.
-func (r *RegionConfig) ApplyZoneConfigExtensionForRegionalIn(
+func (r *RegionConfig) ExtendZoneConfigWithRegionalIn(
 	zc zonepb.ZoneConfig, region catpb.RegionName,
-) zonepb.ZoneConfig {
+) (zonepb.ZoneConfig, error) {
+	var numVoters, numReplicas int32
+	if zc.NumVoters != nil {
+		numVoters = *zc.NumVoters
+	}
+	if zc.NumReplicas != nil {
+		numReplicas = *zc.NumReplicas
+	}
+
 	if ext := r.zoneCfgExtensions.Regional; ext != nil {
+		if ext.LeasePreferences != nil {
+			return zonepb.ZoneConfig{}, errors.New("REGIONAL zone config extensions " +
+				"are not allowed to set lease_preferences")
+		}
 		zc = extendZoneCfg(zc, *ext)
 	}
 	if ext, ok := r.zoneCfgExtensions.RegionalIn[region]; ok {
+		if err := validateZoneConfigExtension(ext, zc, region.String()); err != nil {
+			return zonepb.ZoneConfig{}, err
+		}
 		zc = extendZoneCfg(zc, ext)
 	}
-	return zc
+
+	// TODO(janexing): to ensure that the zone config extension won't break the
+	// survival goal, here we only add restriction on the number of voters and
+	// replicas. We may want to consider adding constraint to their distribution
+	// across zones/regions as well.
+	if zc.NumVoters != nil && *zc.NumVoters < numVoters {
+		return zonepb.ZoneConfig{}, errors.Newf("zone config extension "+
+			"cannot set a voter number %v that is lower than the one required for the "+
+			"survival goal: %v with goal %v\n", *zc.NumVoters, numVoters, r.SurvivalGoal())
+	}
+	if zc.NumReplicas != nil && *zc.NumReplicas < numReplicas {
+		return zonepb.ZoneConfig{}, errors.Newf("zone config extension "+
+			"cannot set a replica number %v that is lower than the one required for the "+
```

Comment:
num_replicas just to be explicit about the setting name

Reply:
Done.